### PR TITLE
KAFKA-14589 ConsumerGroupCommand rewritten in java

### PR DESCRIPTION
**Pull request by nizhikov**
This PR contains changes to rewrite `ConsumerGroupCommand` in java and transfer it to `tools` module

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: tools/src/main/java/org/apache/kafka/tools/consumergroup/CsvUtils.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        else
+            throw new IllegalStateException("Unhandled class " + clazz);
+
+        return mapper.schemaFor(clazz).sortedBy(fields);
+    }
+
+    public interface CsvRecord {
+    }
+
+    public static class CsvRecordWithGroup implements CsvRecord {
```
Comment (tledkov) [inline tools/src/main/java/org/apache/kafka/tools/consumergroup/CsvUtils.java]: What about `CsvRecordWithGroup extends CsvRecordNoGroup`?
Also looks like setters for class properties can be removed because they are not used by code and class data members have jackson annotations.

Comment (nizhikov) [reply tools/src/main/java/org/apache/kafka/tools/consumergroup/CsvUtils.java]: fixed. Thanks

**Comment by jolshan** [comment]
I have noticed several 24hr+ builds for the project when we should time out after 8. I've alerted the mailing thread, but will try to do some of my own investigation.

**Comment by nizhikov** [comment]
Hello @mimaison @jolshan @showuon 

I open two PR's to move the rest of `ConsoleGroupCommand` tests to tools.

#15365 - pretty small (+200, -100)
#15363 - relatively big, but contains only two test classes.

Please, take a look.

**Comment by chia7712** [comment]
@nizhikov thanks for taking over this big issue. I'd like to give a review. However, it is a bit huge. Is it possible to split this PR into small PRs by the "command" of `ConsumerGroupCommand`? For example, we can have the new `ConsumerGroupCommand` supporting only `list`, and then `describe`, `delete`, etc. And the last PR add deprecation information and cleanup. WDYT?

**Comment by nizhikov** [comment]
Hello @chia7712 

Yes.

There are two separate PR's that moves tests of command to `tools`:

https://github.com/apache/kafka/pull/15365 - pretty small (+200, -100)
https://github.com/apache/kafka/pull/15363 - relatively big, but contains only two test classes.

Please, take a look.

**Comment by nizhikov** [comment]
> For example, we can have the new ConsumerGroupCommand supporting only list, and then describe, delete, etc. And the last PR add deprecation information and cleanup. WDYT?

I think we can decide on moving command step by step after all tests will be moved to `tools`. 
From my experience with previous command - we can move whole command after all tests are moved.

But, if patch size will remain too big - I will follow your advice and prepare PR's to move one mode at a time.

**Comment by chia7712** [comment]
> There are two separate PR's that moves tests of command to tools:

oh, sorry that I did not notice that you have tried to make small PRs. will take a look later.

> But, if patch size will remain too big - I will follow your advice and prepare PR's to move one mode at a time.

💯

**Comment by nizhikov** [comment]
Hello @mimaison @jolshan @showuon

I split PR even more and now all three PR's for a command test seems pretty small.
Please, take a look if you have spare cycles.

#15365
#15363
#15465

**Comment by chia7712** [comment]
@nizhikov please rebase it. I feel we can complete this hard work this week :)

**Comment by nizhikov** [comment]
@chia7712 PR merged with latest trunk and ready for review. Please, take a look.

**Comment by nizhikov** [comment]
@chia7712 CI looks OK for me. Please, take a look.

File: build.gradle [hunk 2]
```diff
@@ -1999,10 +1999,14 @@ project(':tools') {
     implementation project(':clients')
     implementation project(':storage')
     implementation project(':server-common')
+    implementation project(':server')
```
Comment (chia7712) [inline build.gradle]: Is this dependency necessary? we don't expect tool module depends on core module.

Comment (nizhikov) [reply build.gradle]: Great catch. Thanks!
It seems this dependency not required. Removed.

**Comment by nizhikov** [comment]
@chia7712 

I ran system tests for command, also.
System tests passed.
```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.11.4
session_id:       2024-03-12--004
run time:         5 minutes 22.467 seconds
tests run:        12
passed:           12
flaky:            0
failed:           0
ignored:          0
================================================================================
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_describe_consumer_group.security_protocol=PLAINTEXT.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=False
status:     PASS
run time:   26.101 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_describe_consumer_group.security_protocol=PLAINTEXT.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=True
status:     PASS
run time:   25.893 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_describe_consumer_group.security_protocol=PLAINTEXT.metadata_quorum=ZK.use_new_coordinator=False
status:     PASS
run time:   21.294 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_describe_consumer_group.security_protocol=SSL.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=False
status:     PASS
run time:   31.293 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_describe_consumer_group.security_protocol=SSL.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=True
status:     PASS
run time:   31.091 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_describe_consumer_group.security_protocol=SSL.metadata_quorum=ZK.use_new_coordinator=False
status:     PASS
run time:   25.902 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_list_consumer_groups.security_protocol=PLAINTEXT.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=False
status:     PASS
run time:   25.983 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_list_consumer_groups.security_protocol=PLAINTEXT.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=True
status:     PASS
run time:   26.154 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_list_consumer_groups.security_protocol=PLAINTEXT.metadata_quorum=ZK.use_new_coordinator=False
status:     PASS
run time:   20.705 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_list_consumer_groups.security_protocol=SSL.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=False
status:     PASS
run time:   30.682 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_list_consumer_groups.security_protocol=SSL.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=True
status:     PASS
run time:   30.673 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_list_consumer_groups.security_protocol=SSL.metadata_quorum=ZK.use_new_coordinator=False
status:     PASS
run time:   26.424 seconds
--------------------------------------------------------------------------------
nizhikov@v10792:~/kafka$ git status
On branch KAFKA-14589
Your branch is up to date with 'origin/KAFKA-14589'.

nothing to commit, working tree clean
nizhikov@v10792:~/kafka$ git remote -v
origin	https://github.com/nizhikov/kafka.git (fetch)
origin	https://github.com/nizhikov/kafka.git (push)
```

File: tools/src/main/java/org/apache/kafka/tools/ToolsUtils.java [hunk 3]
```diff
@@ -155,4 +157,17 @@ public static <T> Set<T> minus(Set<T> set, T...toRemove) {
         return res;
     }
 
+    /**
+     * This is a simple wrapper around `CommandLineUtils.printUsageAndExit`.
+     * It is needed for tools migration (KAFKA-14525), as there is no Java equivalent for return type `Nothing`.
+     * Can be removed once [[kafka.admin.ConsumerGroupCommand]], [[kafka.tools.ConsoleConsumer]]
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/ToolsUtils.java]: `kafka.admin.ConsumerGroupCommand` is migrated if this PR gets merged, right?

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java [hunk 4]
```diff
@@ -0,0 +1,1241 @@
...
+            TopicPartition topicPartition = new TopicPartition(topic, partition);
+            OffsetAndMetadata offsetAndMetadata = new OffsetAndMetadata(offset);
+            Map<TopicPartition, OffsetAndMetadata> dataMap = acc.getOrDefault(group, new HashMap<>());
+            dataMap.put(topicPartition, offsetAndMetadata);
+            acc.put(group, dataMap);
+            return acc;
+        }
+
+        private Map<String, Map<TopicPartition, OffsetAndMetadata>> parseResetPlan(String resetPlanCsv) {
+            ObjectReader csvReader = new CsvUtils().readerFor(CsvUtils.CsvRecordNoGroup.class);
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]: We should not instantiate a `utils` class. Maybe the method can be a static method?

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        else
+            throw new IllegalStateException("Unhandled class " + clazz);
+
+        return mapper.schemaFor(clazz).sortedBy(fields);
+    }
+
+    public interface CsvRecord {
+    }
+
+    public static class CsvRecordWithGroup extends CsvRecordNoGroup {
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java]: this naming is weird to me. It should be fine to de-couple them by adding all fields to `CsvRecordWithGroup`

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            fields = CsvRecordWithGroup.FIELDS;
+        else if (CsvRecordNoGroup.class == clazz)
+            fields = CsvRecordNoGroup.FIELDS;
+        else
+            throw new IllegalStateException("Unhandled class " + clazz);
+
+        return mapper.schemaFor(clazz).sortedBy(fields);
+    }
+
+    public interface CsvRecord {
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java]: Do we need this dumb interface? We use `ObjectWriter` to output strings and it can take `Object` directly.

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/GroupState.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.tools.consumer.group;
+
+import org.apache.kafka.common.Node;
+
+class GroupState {
+    public final String group;
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/GroupState.java]: Can we narrow the scope from `public` to `package-private`?

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java [hunk 8]
```diff
@@ -0,0 +1,1241 @@
...
+
+public class ConsumerGroupCommand {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConsumerGroupCommand.class);
+
+    public static final String MISSING_COLUMN_VALUE = "-";
+
+    public static void main(String[] args) {
+        ConsumerGroupCommandOptions opts = new ConsumerGroupCommandOptions(args);
+        try {
+            opts.checkArgs();
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]: It seems to me we should instantiate the object only if it is already checked and validated. Hence, could we change the code style:
```java
ConsumerGroupCommandOptions opts = ConsumerGroupCommandOptions.fromArgs(args);
```

and `ConsumerGroupCommandOptions` should have constructor taking all fields instead of "args". the helper `ConsumerGroupCommandOptions.fromArgs` does parse and check the args

File: bin/kafka-consumer-groups.sh [hunk 9]
```diff
@@ -14,4 +14,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-exec $(dirname $0)/kafka-run-class.sh kafka.admin.ConsumerGroupCommand "$@"
+exec $(dirname $0)/kafka-run-class.sh org.apache.kafka.tools.consumer.group.ConsumerGroupCommand "$@"
```
Comment (chia7712) [inline bin/kafka-consumer-groups.sh:17]: Could you please test it manually?

**Review by chia7712 — COMMENTED** [review]
@nizhikov sorry for late review. This is the last one and it is good time to give more reviews. please take a look at comments, thanks!!

File: tools/src/main/java/org/apache/kafka/tools/ToolsUtils.java (hunk 3 above)
Comment (nizhikov) [reply tools/src/main/java/org/apache/kafka/tools/ToolsUtils.java]: Yes

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java (hunk 5 above)
Comment (nizhikov) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java]: Done.

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java (hunk 6 above)
Comment (nizhikov) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java]: Done

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/GroupState.java (hunk 7 above)
Comment (nizhikov) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/GroupState.java]: visibility minified for all new classes

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java (hunk 4 above)
Comment (nizhikov) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]: Done

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java (hunk 8 above)
Comment (nizhikov) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]: Done

File: bin/kafka-consumer-groups.sh (hunk 9 above)
Comment (nizhikov) [reply bin/kafka-consumer-groups.sh:17]: Basic manual testing done.


```
❯ ./bin/kafka-consumer-groups.sh --bootstrap-server localhost:9092 --describe --all-groups

Consumer group 'console-consumer-79292' has no active members.

GROUP           TOPIC           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             CONSUMER-ID                                           HOST            CLIENT-ID
my_group        first_topic     0          0               0               0               console-consumer-9db50227-1820-46a6-b667-a387553ac536 /192.168.31.66  console-consumer
my_group        first_topic     1          11              11              0               console-consumer-9db50227-1820-46a6-b667-a387553ac536 /192.168.31.66  console-consumer
my_group        first_topic     2          0               0               0               console-consumer-9db50227-1820-46a6-b667-a387553ac536 /192.168.31.66  console-consumer%
❯ ./bin/kafka-consumer-groups.sh --bootstrap-server localhost:9092 --list
console-consumer-79292
my_group
```

**Comment by nizhikov** [comment]
@chia7712 

I addressed all your comments. Please, take a look one more time.

> This is the last one and it is good time to give more reviews

Are you suggesting to find more reviewers for this PR?

**Comment by chia7712** [comment]
> Are you suggesting to find more reviewers for this PR?

Yep, if they have free cycle 😃 

Anyway, I will take a look later. Thanks for updating code

**Comment by nizhikov** [comment]
Hello @mimaison , @jolshan 

Do you have time to review this PR?
We are ready to move `ConsumerGroupCommand` to tools - all tests moved to java, I've checked system tests and check java version manually.

**Comment by chia7712** [comment]
@nizhikov sorry for late review. could you rebase code and I will take a look later.

**Comment by nizhikov** [comment]
@chia7712 PR merged with the latest trunk.

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java [hunk 10]
```diff
@@ -0,0 +1,1240 @@
...
+                    if (offsetsResultInfo.offset() != ListOffsetsResponse.UNKNOWN_OFFSET)
+                        successfulOffsetsForTimes.put(tp, offsetsResultInfo);
+                    else
+                        unsuccessfulOffsetsForTimes.put(tp, offsetsResultInfo);
+                });
+
+                Map<TopicPartition, LogOffsetResult> successfulLogTimestampOffsets = successfulOffsetsForTimes.entrySet().stream()
+                    .collect(Collectors.toMap(Map.Entry::getKey, e -> new LogOffset(e.getValue().offset())));
+
+                unsuccessfulOffsetsForTimes.forEach((tp, offsetResultInfo) -> {
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]: please remove the `{`

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java [hunk 11]
```diff
@@ -0,0 +1,1240 @@
...
+                return adminClient.listConsumerGroupOffsets(
+                    Collections.singletonMap(groupId, new ListConsumerGroupOffsetsSpec()),
+                    withTimeoutMs(new ListConsumerGroupOffsetsOptions())
+                ).partitionsToOffsetAndMetadata(groupId).get();
+            } catch (InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        private Map<String, Map<TopicPartition, OffsetAndMetadata>> updateGroupMetadata(String group, String topic, int partition, long offset, Map<String, Map<TopicPartition, OffsetAndMetadata>> acc) {
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]: this method is unused. Is it useless in scala version before?

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java [hunk 12]
```diff
@@ -0,0 +1,1240 @@
...
+        }
+
+        private void printOffsets(Map<String, Tuple2<Optional<String>, Optional<Collection<PartitionAssignmentState>>>> offsets) {
+            offsets.forEach((groupId, tuple) -> {
+                Optional<String> state = tuple.v1;
+                Optional<Collection<PartitionAssignmentState>> assignments = tuple.v2;
+
+                if (shouldPrintMemberState(groupId, state, size(assignments))) {
+                    // find proper columns width
+                    int maxGroupLen = 15, maxTopicLen = 15, maxConsumerIdLen = 15, maxHostLen = 15;
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]: Could we move those code to a static method?

File: core/src/test/scala/unit/kafka/utils/TestUtils.scala [hunk 13]
```diff
@@ -664,7 +664,7 @@ object TestUtils extends Logging {
    * until the leader is elected and metadata is propagated to all brokers. If it does, the method verifies that it has
    * the expected number of partition and replication factor however it does not guarantee that the topic is empty.
    */
-  def createOffsetsTopic(zkClient: KafkaZkClient, servers: Seq[KafkaBroker]): Unit = {
+  def createOffsetsTopic[B <: KafkaBroker](zkClient: KafkaZkClient, servers: Seq[B]): Unit = {
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/utils/TestUtils.scala]: why we need this change?

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java [hunk 14]
```diff
@@ -144,18 +145,18 @@ private void testWithConsumerGroup(java.util.function.Consumer<Runnable> withCon
         withConsumerGroup.accept(() -> {
             String topic = inputPartition >= 0 ? inputTopic + ":" + inputPartition : inputTopic;
             ConsumerGroupCommand.ConsumerGroupService service = getConsumerGroupService(getArgs(GROUP, topic));
-            scala.Tuple2<Errors, scala.collection.Map<TopicPartition, Throwable>> res = service.deleteOffsets(GROUP, seq(Collections.singletonList(topic)).toList());
-            Errors topLevelError = res._1;
-            scala.collection.Map<TopicPartition, Throwable> partitions = res._2;
+            Tuple2<Errors, Map<TopicPartition, Throwable>> res = service.deleteOffsets(GROUP, Collections.singletonList(topic));
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java:148]: (this is unrelated to this PR)

Could we re-use the `Entry` to replace `Tuple2`? That is a java standard API and we can create a immutable impl by calling `new AbstractMap.SimpleImmutableEntry`, and it can be simplified to `Map.entry` after we get rid of JDK8

Comment (chia7712) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java:148]: BTW, we can address that in a separate PR if the comment is valid.

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+        @JsonProperty
+        private String topic;
+
+        @JsonProperty
+        private int partition;
+
+        @JsonProperty
+        private long offset;
+
+        public CsvRecordNoGroup() {
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java:124]: Could you add comment to explain that the public constructor is necessary to jackson? otherwise, it looks like a useless constructor :)

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        @JsonProperty
+        private String topic;
+
+        @JsonProperty
+        private int partition;
+
+        @JsonProperty
+        private long offset;
+
+        public CsvRecordWithGroup() {
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java:66]: ditto

File: checkstyle/import-control.xml [hunk 17]
```diff
@@ -330,6 +330,7 @@
         <allow pkg="kafka.zk" />
         <allow pkg="org.apache.kafka.tools"/>
         <allow pkg="org.apache.kafka.server.config" />
+        <allow pkg="org.apache.kafka.metadata.authorizer"/>
```
Comment (chia7712) [inline checkstyle/import-control.xml]: Is this necessary? I remove it and build still works well.

**Review by chia7712 — COMMENTED** [review]
@nizhikov thanks for this patch. I left a couple of comments. please take a look

File: core/src/test/scala/unit/kafka/utils/TestUtils.scala (hunk 13 above)
Comment (nizhikov) [reply core/src/test/scala/unit/kafka/utils/TestUtils.scala]: There was compilation error earlier.
Looks like now it's gone.
Change reverted.

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java (hunk 11 above)
Comment (nizhikov) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]: Unused method removed.

In the java version I just inline method logic 
looks like it pretty simple and can be inlined.

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java (hunk 12 above)
Comment (nizhikov) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]: Done.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java (hunk 14 above)
Comment (nizhikov) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java:148]: I think usage of Entry is a great idea. 
Will implement it after this PR.

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java (hunk 15 above)
Comment (nizhikov) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java:124]: Done

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java (hunk 16 above)
Comment (nizhikov) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java:66]: Done

File: checkstyle/import-control.xml (hunk 17 above)
Comment (nizhikov) [reply checkstyle/import-control.xml]: This not required.

Sorry for that, should been check this before review.

**Comment by nizhikov** [comment]
@chia7712 I addressed all your comments. Please, take a look one more time.

**Comment by chia7712** [comment]
the failed test pass on my local:
```sh
./gradlew cleanTest :streams:test --tests HighAvailabilityTaskAssignorIntegrationTest.shouldScaleOutWithWarmupTasksAndPersistentStores :tools:test --tests MetadataQuorumCommandTest.testDescribeQuorumStatusSuccessful :metadata:test --tests QuorumControllerTest.testBrokerHeartbeatDuringMigration :core:test --tests PlaintextAdminIntegrationTest.testReplicaCanFetchFromLogStartOffsetAfterDeleteRecords --tests ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric --tests LogDirFailureTest.testIOExceptionDuringLogRoll --tests KafkaZkClientTest.testConcurrentKRaftControllerClaim --tests ZkMigrationIntegrationTest.testMigrateTopicDeletions --tests ReplicationQuotasTest.shouldBootstrapTwoBrokersWithFollowerThrottle
```
Also, the system test `consumer_group_command_test.py` pass on my local too. will merge it later

**Review by chia7712 — APPROVED** [review]

**Comment by chia7712** [comment]
@nizhikov thanks for all efforts! Please file PR to address follow-up, thanks!

**Comment by nizhikov** [comment]
@chia7712 I tried to perform Tuple2 -> Map.Entry substitution.

The only issue I can see is naming.
`Map.Entry` has `getKey`, `getValue` for first and second values in pair which  sets some kind of relation between values for me.
While Tuple2 just stores two values(v1, v2) that sits together in some piece of code.

Should we go with `getKey`, `getValue` for tuple values? Is it clear naming? WDYT?

**Comment by nizhikov** [comment]
@chia7712 
@jolshan 
@mimaison 
@tledkov 

Guys, thank you for a help with this PR!

**Comment by chia7712** [comment]
> The only issue I can see is naming.
Map.Entry has getKey, getValue for first and second values in pair which sets some kind of relation between values for me.
While Tuple2 just stores two values(v1, v2) that sits together in some piece of code.
Should we go with getKey, getValue for tuple values? Is it clear naming? WDYT?

My point was that we can leverage the JAVA API instead of creating new one :)

**Comment by nizhikov** [comment]
@chia7712 Please, take a look - https://github.com/apache/kafka/pull/15560
